redhat-cop/infra-ansible
# Adding a new module/role: fetch_user_data_gsheets

### What does this PR do?
This PR adds a new module/role called `fetch_user_data_gsheets` which is used to pull data from a google spreadsheet and convert it into json to be used by other roles (such as manage-users).

A spreadsheet like the following is useful for data-gathering purposes since it is user friendly and easy to maintain. With this module it can be easy to automate the changes to a spreadsheet into a system like IdM.

| user_name | first_name | last_name | email                 | expiration_date | group      |
|-----------|------------|-----------|-----------------------|-----------------|------------|
| brock.lee     | Brock      | Lee       | brock.lee@example.com |                 | core_users |

The module will assign data to each user based on the headers in the first row (order doesn't matter). 

### How should this be tested?
Create a google sheet which looks like the above example. The only two mandatory headers are email and group right now (group is required since it is not only outputting a users json object, but also a user_groups json object).
Follow the instructions from the README (copied below) to create a google service account to connect to the spreadsheet.

1. Go to the [Google Developer Console](https://console.developers.google.com/project).
2. Go to the credentials page.
3. Press `Create Credentials` -> `Service account key`.
4. Select `New service account`.
5. Set the Role to `Project -> Editor`. Set the account name and account ID as desired.
6. Create the key with type `json`.
7. Save this .json file to your desired location (the path will need to be referenced later). Note: If storing with your code in a repo, add this file to the `.gitignore`.
8. In the credentials file, find the `client_email` key and copy the matching value. This email is the service account.
9. Add that service account email as an editor into your Google sheet.

Now you can run the module with a playbook like the following:
```yml
- hosts: localhost
  vars:
    credentials_src: /path/to/credentials.json
    spreadsheet_url: https://docs.google.com/spreadsheets/d/google_sheet_extension
    worksheet_name: Sheet1
  roles:
    - fetch-data-gsheets
    - json-to-idm
    - json-to-atlassian
```
If the credentials are working properly, you should have a successful run. These roles will print out the json and variables so you can verify it is outputting as expected. If you want to test fully end-to-end you can add these to the roles list to show that it correctly connects with the manage roles:
```yaml
    - manage-users
    - manage-atlassian-users
```
Along with the variables for an IdM and an Atlassian service:
```yaml
    ipa_admin_user: admin
    ipa_admin_password: 'my_password'
    ipa_host: idm.example.com
    atlassian_url: https://test.atlassian.net
    atlassian_username: my_user
    atlassian_password: "my_password"
```

### Is there a relevant Issue open for this?
N/A

### Other Relevant info, PRs, etc.
Please provide link to other PRs that may be related (blocking, resolves, etc. etc.)

### People to notify
cc: @redhat-cop/infra-ansible   @oybed @day4skiing
